frontend/catalyst/pennylane_extensions.py

```diff
@@ -168,54 +187,108 @@ class QJITDevice(qml.QubitDevice):
-        "Identity",
-        "S",
-        "T",
-        "PhaseShift",
-        "RX",
-        "RY",
-        "RZ",
-        "CNOT",
-        "CY",
-        "CZ",
-        "SWAP",
-        "IsingXX",
-        "IsingYY",
-        "IsingXY",
-        "IsingZZ",
-        "ControlledPhaseShift",
-        "CRX",
-        "CRY",
-        "CRZ",
-        "CRot",
-        "CSWAP",
-        "MultiRZ",
-        "QubitUnitary",
-        "Adjoint",
-    ]
-    observables = [
-        "Identity",
-        "PauliX",
-        "PauliY",
-        "PauliZ",
-        "Hadamard",
-        "Hermitian",
-        "Hamiltonian",
-    ]
+
+    # These must be present even if empty.
+    operations = []
+    observables = []
+
+    @staticmethod
+    def _get_operations_available_in_QIR_abstract_machine():
+        # These are the methods implemented in the runtime.
+        # A device might have gates which are native to the device
+        # but are not exposed. This is a limitation of the current runtime
+        # because it does not contain implementation of gates for every possible
+        # gate from the infinite set.
+        return {
+            "Identity",
+            "PauliX",
+            "PauliY",
+            "PauliZ",
+            "Hadamard",
+            "S",
+            "T",
+            "PhaseShift",
+            "RX",
+            "RY",
+            "RZ",
+            "Rot",
+            "CNOT",
+            "CY",
+            "CZ",
+            "SWAP",
+            "IsingXX",
+            "IsingYY",
+            "IsingXY",
+            "ControlledPhaseShift",
+            "CRX",
+            "CRY",
+            "CRZ",
+            "CRot",
+            "CSWAP",
+            "Toffoli",
+            "MultiRZ",
+            "QubitUnitary",
+        }
+
+    @staticmethod
+    def _get_operations_to_convert_to_matrix(_config):
+        # We currently override and only set MultiControlledX to preserve current behaviour.
+        # We could choose to read from config and use the "matrix" gates.
+        # However, that affects differentiability.
+        # None of the "matrix" gates with more than 2 qubits parameters are differentiable.
+        # TODO: https://github.com/PennyLaneAI/catalyst/issues/398
+        return {"MultiControlledX"}
+
+    @staticmethod
+    def _check_mid_circuit_measurement(config):
+        return config["compilation"]["mid_circuit_measurement"]
+
+    @staticmethod
+    def _check_adjoint(config):
+        return config["compilation"]["quantum_adjoint"]
+
+    @staticmethod
+    def _check_quantum_control(config):
+        return config["compilation"]["quantum_control"]
+
+    @staticmethod
+    def _set_supported_operations(config):
+        """Override the set of supported operations."""
+        native_gates = set(config["operators"]["gates"][0]["native"])
+        qir_gates = QJITDevice._get_operations_available_in_QIR_abstract_machine()
+        QJITDevice.operations = list(native_gates.intersection(qir_gates))
```

Comment:
The runtime at the moment has a design issue. For every "native" gate in the device there must be an implementation of a QIR quantum instruction set function. So if a gate specifies a native gate "Foo" and we don't implement "foo" in the runtime, the compiler will generate a function call to `__quantum__qis__Foo` but the linker will fail to find an implementation for it at link time. That's why we have the intersection here. That way if "Foo" is native, but it is not available as a `__quantum__qis__Foo` function, then we don't generate a call to `__quantum__qis__Foo` and avoid the linking error.

This can be solved in different ways and we can chat about this during a meeting or offline.

I had a previously suggested that the meaning of the TOML file should change to avoid this intersection and respect the contents of the TOML file. See [here](https://github.com/PennyLaneAI/pennylane-lightning/pull/566#discussion_r1412189268)

Reply:
Happy to continue the discussion here or offline.

Reply:
I'll close this comment if there are no further questions.